src/Telemetry/AzureDevOpsClient/AzureDevOpsClient.cs

```diff
@@ -155,25 +156,43 @@ public async Task<WorkItem> CreateRcaWorkItem(string project, string title, Canc
         return JsonConvert.DeserializeObject<WorkItem>(json);
     }
 
+    /// <summary>
+    /// The method reads the logs as a stream, line by line and tries to match the regexes in order. 
+    /// If the regexes are matched in order, the last match is returned.
+    /// </summary>
     public async Task<string> TryGetImageName(
         string logUri,
-        Func<string, string> findImageName,
+        Regex[] regexes,
         CancellationToken cancellationToken)
     {
-
         using var request = new HttpRequestMessage(HttpMethod.Get, logUri);
         request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/plain"));
 
         using var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
         using Stream logStream = await response.Content.ReadAsStreamAsync(cancellationToken);
         using StreamReader reader = new StreamReader(logStream);
+        int regexIndex = 0;
         string line;
-        while ((line = reader.ReadLine()) != null)
+        while ((line = await reader.ReadLineAsync()) != null)
         {
-            var imageName = findImageName(line);
-            if (imageName != null)
+            if (TryMatchRegex(line, regexes[regexIndex], out string imageName))
             {
-                return imageName;
+                regexIndex++;
+
+                if (regexIndex == regexes.Length)
+                {
+                    return imageName;
+                }
```

Comment:
nit: My personal preference would be to not modify `regexIndex` as it gets confusing why we change the iterator like that. We also do not zero it, so if the next regex would match the next line, we would get an interesting result here.

I'd say it's better to do:
```csharp
if (regexIndex == regexes.Length - 1)
```

Reply:
We have to modigy the `regexIndex` to iterate over the regexes. We're basically searching for the consecutive lines that whill all have a match in the regexes we provide, we zero the `regexIndex` in case of a miss so we can start over

Reply:
Oh, so for 3 regexes we should match 3 different lines each with its own regex?

Reply:
yes, we currently have a case where for OneES images we need to match two lines in a row. We'll be able to use this in case we need to match more lines in the future

Reply:
What if we have these lines:

```
a
b
b
c
```

and these regexes:
```
a|b
a|b
c
```

We will match the first two lines, fail at the third:

```
a ✅ a|b
b ✅ a|b
b ❌ c
c
```

Start over at the third line and fail to match this:

```
b ✅ a|b
c ❌ a|b
```

While obviously the regexes match the last three lines of the input

Reply:
This is true, this is such an interview question ahah

Reply:
I think I fixed the algorithm, I will write some tests to verify it